src/memory/memory_control.cpp

```diff
@@ -10,18 +10,297 @@
+  } else {
+    memgraph::utils::total_memory_tracker.FreeVirt(static_cast<int64_t>(size));
+  }
+
+  old_hooks->destroy(extent_hooks, addr, size, committed, arena_ind);
+}
+
+static bool my_commit(extent_hooks_t *extent_hooks, void *addr, size_t size, size_t offset, size_t length,
+                      unsigned arena_ind) {
+  auto err = old_hooks->commit(extent_hooks, addr, size, offset, length, arena_ind);
+
+  if (err) {
+    return err;
+  }
+
+  memgraph::utils::total_memory_tracker.FreeVirt(static_cast<int64_t>(length));
+  memgraph::utils::total_memory_tracker.Alloc(static_cast<int64_t>(length));
+
+  return false;
+}
+
+static bool my_decommit(extent_hooks_t *extent_hooks, void *addr, size_t size, size_t offset, size_t length,
+                        unsigned arena_ind) {
+  MG_ASSERT(old_hooks && old_hooks->decommit);
+  auto err = old_hooks->decommit(extent_hooks, addr, size, offset, length, arena_ind);
+
+  if (err) {
+    return err;
+  }
+
+  memgraph::utils::total_memory_tracker.AllocVirt(static_cast<int64_t>(length));
+  memgraph::utils::total_memory_tracker.Free(static_cast<int64_t>(length));
+
+  return false;
+}
+
+static bool my_purge_forced(extent_hooks_t *extent_hooks, void *addr, size_t size, size_t offset, size_t length,
+                            unsigned arena_ind) {
+  MG_ASSERT(old_hooks && old_hooks->purge_forced);
+  auto err = old_hooks->purge_forced(extent_hooks, addr, size, offset, length, arena_ind);
+
+  if (err) {
+    return err;
+  }
+  memgraph::utils::total_memory_tracker.Free(static_cast<int64_t>(length));
+
+  return false;
+}
+
+static bool my_purge_lazy(extent_hooks_t *extent_hooks, void *addr, size_t size, size_t offset, size_t length,
+                          unsigned arena_ind) {
+  // If memory is purged lazily, it will not be cleaned immediatelly if we are not using MADVISE_DONTNEED (muzzy=0 and
+  // decay=0)
+  MG_ASSERT(old_hooks && old_hooks->purge_lazy);
+  return old_hooks->purge_lazy(extent_hooks, addr, size, offset, length, arena_ind);
+}
+
+static bool my_split(extent_hooks_t *extent_hooks, void *addr, size_t size, size_t size_a, size_t size_b,
+                     bool committed, unsigned arena_ind) {
+  MG_ASSERT(old_hooks && old_hooks->split);
+  return old_hooks->split(extent_hooks, addr, size, size_a, size_b, committed, arena_ind);
+}
+
+static bool my_merge(extent_hooks_t *extent_hooks, void *addr_a, size_t size_a, void *addr_b, size_t size_b,
+                     bool committed, unsigned arena_ind) {
+  MG_ASSERT(old_hooks && old_hooks->merge);
+  return old_hooks->merge(extent_hooks, addr_a, size_a, addr_b, size_b, committed, arena_ind);
+}
+
+static constexpr extent_hooks_t custom_hooks = {
+    .alloc = &my_alloc,
+    .dalloc = &my_dalloc,
+    .destroy = &my_destroy,
+    .commit = &my_commit,
+    .decommit = &my_decommit,
+    .purge_lazy = &my_purge_lazy,
+    .purge_forced = &my_purge_forced,
+    .split = &my_split,
+    .merge = &my_merge,
```

Comment:
Why not just use old_hooks directly...why add runtime assert + indirection

Reply:
Hmm what do you mean?

Reply:
I need to use different hooks, because if I use old hooks then jemalloc won't call my hooks and I can't count memory.

Reply:
But these hooks don't do any counting, they assert + pass to old hooks

Reply:
Hmm, they do. If you check alloc, purge_forced, dalloc, there are calls to memory tracker?

Reply:
Ah, you are thinking on only few of them, like merge and split, not purge_forced also. Yes, that makes sense